
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ericmjl.github.io/fast-seqfunc/tutorials/regression_tutorial/">
      
      
        <link rel="prev" href="../classification_tutorial/">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.9">
    
    
      
        <title>Sequence Regression with Fast-SeqFunc - fast-seqfunc</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4af4bdda.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../apidocs.css">
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#sequence-regression-with-fast-seqfunc" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="fast-seqfunc" class="md-header__button md-logo" aria-label="fast-seqfunc" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M96 0C43 0 0 43 0 96v320c0 53 43 96 96 96h320c17.7 0 32-14.3 32-32s-14.3-32-32-32v-64c17.7 0 32-14.3 32-32V32c0-17.7-14.3-32-32-32H96m0 384h256v64H96c-17.7 0-32-14.3-32-32s14.3-32 32-32m32-240c0-8.8 7.2-16 16-16h192c8.8 0 16 7.2 16 16s-7.2 16-16 16H144c-8.8 0-16-7.2-16-16m16 48h192c8.8 0 16 7.2 16 16s-7.2 16-16 16H144c-8.8 0-16-7.2-16-16s7.2-16 16-16"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            fast-seqfunc
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Sequence Regression with Fast-SeqFunc
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/ericmjl/fast-seqfunc" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    ericmjl/fast-seqfunc
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="fast-seqfunc" class="md-nav__button md-logo" aria-label="fast-seqfunc" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M96 0C43 0 0 43 0 96v320c0 53 43 96 96 96h320c17.7 0 32-14.3 32-32s-14.3-32-32-32v-64c17.7 0 32-14.3 32-32V32c0-17.7-14.3-32-32-32H96m0 384h256v64H96c-17.7 0-32-14.3-32-32s14.3-32 32-32m32-240c0-8.8 7.2-16 16-16h192c8.8 0 16 7.2 16 16s-7.2 16-16 16H144c-8.8 0-16-7.2-16-16m16 48h192c8.8 0 16 7.2 16 16s-7.2 16-16 16H144c-8.8 0-16-7.2-16-16s7.2-16 16-16"/></svg>

    </a>
    fast-seqfunc
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/ericmjl/fast-seqfunc" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    ericmjl/fast-seqfunc
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fast-SeqFunc Documentation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Top-level API for fast-seqfunc
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fast-SeqFunc Quickstart
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../roadmap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Roadmap
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Design
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Design
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../design/additional_predictors/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional Predictors in Fast-SeqFunc: Design Document
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../design/additional_predictors_example/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional Predictors: Usage Examples
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../design/custom_alphabets/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Custom Alphabets Design Document
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../design/overall/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fast-SeqFunc: Design Document
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Releases
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Releases
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../releases/v0.0.3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    V0.0.3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../releases/v0.1.0/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    V0.1.0
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../releases/v0.1.1/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    V0.1.1
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" checked>
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../classification_tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Sequence Classification with Fast-SeqFunc
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Sequence Regression with Fast-SeqFunc
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Sequence Regression with Fast-SeqFunc
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      Prerequisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-sequence-function-data" class="md-nav__link">
    <span class="md-ellipsis">
      Working with Sequence-Function Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Working with Sequence-Function Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparing-your-own-data" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing Your Own Data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#splitting-data-for-training-and-testing" class="md-nav__link">
    <span class="md-ellipsis">
      Splitting Data for Training and Testing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#training-a-regression-model" class="md-nav__link">
    <span class="md-ellipsis">
      Training a Regression Model
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Training a Regression Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#understanding-embedding-methods" class="md-nav__link">
    <span class="md-ellipsis">
      Understanding Embedding Methods
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-predictions" class="md-nav__link">
    <span class="md-ellipsis">
      Making Predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#evaluating-regression-performance" class="md-nav__link">
    <span class="md-ellipsis">
      Evaluating Regression Performance
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interpreting-signal-and-making-decisions" class="md-nav__link">
    <span class="md-ellipsis">
      Interpreting Signal and Making Decisions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Interpreting Signal and Making Decisions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#using-models-for-candidate-ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Using Models for Candidate Ranking
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualizing-results" class="md-nav__link">
    <span class="md-ellipsis">
      Visualizing Results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-different-sequence-types" class="md-nav__link">
    <span class="md-ellipsis">
      Working with Different Sequence Types
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-model-training-options" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced Model Training Options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#loading-and-reusing-models" class="md-nav__link">
    <span class="md-ellipsis">
      Loading and Reusing Models
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      Conclusion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      Prerequisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-sequence-function-data" class="md-nav__link">
    <span class="md-ellipsis">
      Working with Sequence-Function Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Working with Sequence-Function Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preparing-your-own-data" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing Your Own Data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#splitting-data-for-training-and-testing" class="md-nav__link">
    <span class="md-ellipsis">
      Splitting Data for Training and Testing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#training-a-regression-model" class="md-nav__link">
    <span class="md-ellipsis">
      Training a Regression Model
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Training a Regression Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#understanding-embedding-methods" class="md-nav__link">
    <span class="md-ellipsis">
      Understanding Embedding Methods
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-predictions" class="md-nav__link">
    <span class="md-ellipsis">
      Making Predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#evaluating-regression-performance" class="md-nav__link">
    <span class="md-ellipsis">
      Evaluating Regression Performance
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interpreting-signal-and-making-decisions" class="md-nav__link">
    <span class="md-ellipsis">
      Interpreting Signal and Making Decisions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Interpreting Signal and Making Decisions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#using-models-for-candidate-ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Using Models for Candidate Ranking
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualizing-results" class="md-nav__link">
    <span class="md-ellipsis">
      Visualizing Results
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-different-sequence-types" class="md-nav__link">
    <span class="md-ellipsis">
      Working with Different Sequence Types
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-model-training-options" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced Model Training Options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#loading-and-reusing-models" class="md-nav__link">
    <span class="md-ellipsis">
      Loading and Reusing Models
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      Conclusion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next Steps
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="sequence-regression-with-fast-seqfunc">Sequence Regression with Fast-SeqFunc</h1>
<p>This tutorial demonstrates how to use <code>fast-seqfunc</code> for regression problems, where you want to predict continuous values from biological sequences.</p>
<h2 id="overview">Overview</h2>
<p>In sequence regression, we want to learn to predict a continuous value (e.g., binding affinity, enzyme efficiency, or protein stability) from a biological sequence (DNA, RNA, or protein). This tutorial will walk you through:</p>
<ol>
<li>Setting up your environment</li>
<li>Preparing sequence-function data</li>
<li>Training a regression model</li>
<li>Evaluating model performance</li>
<li>Making predictions on new sequences</li>
<li>Visualizing results</li>
</ol>
<h2 id="prerequisites">Prerequisites</h2>
<ul>
<li>Python 3.11 or higher</li>
<li>The following packages:
  <div class="highlight"><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>fast-seqfunc<span class="w"> </span>pandas<span class="w"> </span>numpy<span class="w"> </span>matplotlib<span class="w"> </span>seaborn<span class="w"> </span>scikit-learn<span class="w"> </span>loguru
</code></pre></div></li>
</ul>
<h2 id="setup">Setup</h2>
<p>First, let's import all necessary packages:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">r2_score</span><span class="p">,</span> <span class="n">mean_absolute_error</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">fast_seqfunc</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_model</span><span class="p">,</span> <span class="n">predict</span><span class="p">,</span> <span class="n">save_model</span><span class="p">,</span> <span class="n">load_model</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">loguru</span><span class="w"> </span><span class="kn">import</span> <span class="n">logger</span>
</code></pre></div>
<h2 id="working-with-sequence-function-data">Working with Sequence-Function Data</h2>
<p>Sequence-function data typically consists of biological sequences paired with measurements of a functional property. For this tutorial, we'll create synthetic data:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">fast_seqfunc</span><span class="w"> </span><span class="kn">import</span> <span class="n">generate_dataset_by_task</span>

<span class="c1"># Generate a GC content dataset as an example</span>
<span class="c1"># (where the function is simply the GC content of DNA sequences)</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">generate_dataset_by_task</span><span class="p">(</span>
    <span class="n">task</span><span class="o">=</span><span class="s2">&quot;gc_content&quot;</span><span class="p">,</span>
    <span class="n">count</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>  <span class="c1"># Number of sequences to generate</span>
    <span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>   <span class="c1"># Sequence length</span>
    <span class="n">noise_level</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>  <span class="c1"># Add some noise to make the task more realistic</span>
<span class="p">)</span>

<span class="c1"># Examine the data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data shape: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Target distribution: min=</span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;function&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;max=</span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;function&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, &quot;</span>
      <span class="sa">f</span><span class="s2">&quot;mean=</span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;function&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="preparing-your-own-data">Preparing Your Own Data</h3>
<p>If you have your own data, it should be structured in a DataFrame with at least two columns:
- A column containing the sequences (e.g., "sequence")
- A column containing the target values (e.g., "function")</p>
<p>For example:
<div class="highlight"><pre><span></span><code><span class="c1"># Load your own data</span>
<span class="c1"># data = pd.read_csv(&quot;your_sequence_function_data.csv&quot;)</span>
</code></pre></div></p>
<h2 id="splitting-data-for-training-and-testing">Splitting Data for Training and Testing</h2>
<p>It's important to evaluate your model on data it hasn't seen during training:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Split into train and test sets (80/20 split)</span>
<span class="n">train_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">0.8</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
<span class="n">train_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:</span><span class="n">train_size</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">test_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">train_size</span><span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data split: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">train_data</span><span class="p">)</span><span class="si">}</span><span class="s2"> train, </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">test_data</span><span class="p">)</span><span class="si">}</span><span class="s2"> test samples&quot;</span><span class="p">)</span>

<span class="c1"># Create output directory for results</span>
<span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<h2 id="training-a-regression-model">Training a Regression Model</h2>
<p>Now we can train a regression model using <code>fast-seqfunc</code>:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Train a regression model</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Training regression model...&quot;</span><span class="p">)</span>
<span class="n">model_info</span> <span class="o">=</span> <span class="n">train_model</span><span class="p">(</span>
    <span class="n">train_data</span><span class="o">=</span><span class="n">train_data</span><span class="p">,</span>
    <span class="n">test_data</span><span class="o">=</span><span class="n">test_data</span><span class="p">,</span>
    <span class="n">sequence_col</span><span class="o">=</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span>  <span class="c1"># Column containing sequences</span>
    <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;function&quot;</span><span class="p">,</span>    <span class="c1"># Column containing target values</span>
    <span class="n">embedding_method</span><span class="o">=</span><span class="s2">&quot;one-hot&quot;</span><span class="p">,</span>  <span class="c1"># Method to convert sequences to numerical features</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;regression&quot;</span><span class="p">,</span>     <span class="c1"># Specify regression task</span>
    <span class="n">optimization_metric</span><span class="o">=</span><span class="s2">&quot;r2&quot;</span><span class="p">,</span>    <span class="c1"># Metric to optimize (r2, rmse, mae)</span>
<span class="p">)</span>

<span class="c1"># Display test results</span>
<span class="k">if</span> <span class="n">model_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;test_results&quot;</span><span class="p">):</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Test metrics from training:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">metric</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">model_info</span><span class="p">[</span><span class="s2">&quot;test_results&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Save the model for later use</span>
<span class="n">model_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;regression_model.pkl&quot;</span>
<span class="n">save_model</span><span class="p">(</span><span class="n">model_info</span><span class="p">,</span> <span class="n">model_path</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model saved to </span><span class="si">{</span><span class="n">model_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="understanding-embedding-methods">Understanding Embedding Methods</h3>
<p>The <code>embedding_method</code> parameter determines how sequences are converted to numerical features:</p>
<ul>
<li><code>"one-hot"</code>: Each position in the sequence is encoded as a one-hot vector indicating which amino acid or nucleotide is present.</li>
</ul>
<p>Future versions of <code>fast-seqfunc</code> will include more advanced embedding methods such as ESM2 for proteins and CARP for nucleic acids.</p>
<h2 id="making-predictions">Making Predictions</h2>
<p>After training, you can use your model to make predictions on new sequences:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Generate some new data for prediction</span>
<span class="n">new_data</span> <span class="o">=</span> <span class="n">generate_dataset_by_task</span><span class="p">(</span>
    <span class="n">task</span><span class="o">=</span><span class="s2">&quot;gc_content&quot;</span><span class="p">,</span>
    <span class="n">count</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
    <span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Make predictions</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">model_info</span><span class="p">,</span> <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">])</span>

<span class="c1"># Create results DataFrame</span>
<span class="n">results_df</span> <span class="o">=</span> <span class="n">new_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">results_df</span><span class="p">[</span><span class="s2">&quot;predicted&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">predictions</span>
<span class="n">results_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;regression_predictions.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">results_df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</code></pre></div>
<h2 id="evaluating-regression-performance">Evaluating Regression Performance</h2>
<p>Let's evaluate our model more thoroughly:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Calculate regression metrics</span>
<span class="n">true_values</span> <span class="o">=</span> <span class="n">test_data</span><span class="p">[</span><span class="s2">&quot;function&quot;</span><span class="p">]</span>
<span class="n">predicted_values</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">model_info</span><span class="p">,</span> <span class="n">test_data</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">])</span>

<span class="c1"># Calculate metrics</span>
<span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">true_values</span><span class="p">,</span> <span class="n">predicted_values</span><span class="p">)</span>
<span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>
<span class="n">r2</span> <span class="o">=</span> <span class="n">r2_score</span><span class="p">(</span><span class="n">true_values</span><span class="p">,</span> <span class="n">predicted_values</span><span class="p">)</span>
<span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">true_values</span><span class="p">,</span> <span class="n">predicted_values</span><span class="p">)</span>

<span class="c1"># Print metrics</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Test Set Performance:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  MSE:  </span><span class="si">{</span><span class="n">mse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  RMSE: </span><span class="si">{</span><span class="n">rmse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  R²:   </span><span class="si">{</span><span class="n">r2</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  MAE:  </span><span class="si">{</span><span class="n">mae</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="interpreting-signal-and-making-decisions">Interpreting Signal and Making Decisions</h2>
<p>A critical purpose of Fast-SeqFunc is to determine if there's meaningful signal in your sequence-function data. Let's look at how to interpret these results and make decisions about next steps:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Define a threshold for &quot;meaningful signal&quot;</span>
<span class="n">r2_threshold</span> <span class="o">=</span> <span class="mf">0.1</span>  <span class="c1"># This is a modest threshold for biological data</span>
<span class="n">signal_detected</span> <span class="o">=</span> <span class="n">r2</span> <span class="o">&gt;</span> <span class="n">r2_threshold</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Signal detected in data: </span><span class="si">{</span><span class="n">signal_detected</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">signal_detected</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Recommendations:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  1. Use model for candidate ranking in next experiments&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  2. Consider exploring more advanced embedding methods&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  3. Signal strength (R²=</span><span class="si">{</span><span class="n">r2</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">) suggests </span><span class="si">{</span><span class="s1">&#39;strong&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">r2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mf">0.3</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;moderate&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">r2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mf">0.1</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;weak&#39;</span><span class="si">}</span><span class="s2"> relationship&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Recommendations:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  1. Review experiment design and data quality&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  2. Consider adding more data or different sequence features&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  3. Explore if other modeling approaches detect signal&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="using-models-for-candidate-ranking">Using Models for Candidate Ranking</h3>
<p>If you detected signal, you can use your model to rank new candidate sequences:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Generate or import candidate sequences</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">fast_seqfunc.synthetic</span><span class="w"> </span><span class="kn">import</span> <span class="n">generate_random_sequences</span>

<span class="c1"># Generate 1000 random candidates</span>
<span class="n">candidate_sequences</span> <span class="o">=</span> <span class="n">generate_random_sequences</span><span class="p">(</span>
    <span class="mi">1000</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">alphabet_type</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span>
<span class="p">)</span>

<span class="c1"># Predict their values</span>
<span class="n">candidate_predictions</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">model_info</span><span class="p">,</span> <span class="n">candidate_sequences</span><span class="p">)</span>

<span class="c1"># Create DataFrame for ranking</span>
<span class="n">candidates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="n">candidate_sequences</span><span class="p">,</span>
    <span class="s2">&quot;predicted_value&quot;</span><span class="p">:</span> <span class="n">candidate_predictions</span>
<span class="p">})</span>

<span class="c1"># Sort by predicted value (ascending or descending based on your goal)</span>
<span class="n">candidates_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;predicted_value&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Select top candidates for experimental testing</span>
<span class="n">top_candidates</span> <span class="o">=</span> <span class="n">candidates_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Top 10 candidates for experimental testing:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">top_candidates</span><span class="p">)</span>

<span class="c1"># Save ranked candidates</span>
<span class="n">candidates_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;ranked_candidates.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
<p>This approach allows you to prioritize which sequences to test next in your experiments, potentially saving significant time and resources.</p>
<h2 id="visualizing-results">Visualizing Results</h2>
<p>Visualizations help in understanding model performance:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Create a prediction vs. actual scatter plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">true_values</span><span class="p">,</span> <span class="n">predicted_values</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">true_values</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">true_values</span><span class="o">.</span><span class="n">max</span><span class="p">()],</span>
         <span class="p">[</span><span class="n">true_values</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">true_values</span><span class="o">.</span><span class="n">max</span><span class="p">()],</span>
         <span class="s1">&#39;r--&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Actual Values&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted Values&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Actual vs. Predicted Values&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;regression_scatter_plot.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

<span class="c1"># Plot residuals</span>
<span class="n">residuals</span> <span class="o">=</span> <span class="n">true_values</span> <span class="o">-</span> <span class="n">predicted_values</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">predicted_values</span><span class="p">,</span> <span class="n">residuals</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Predicted Values&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Residuals&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Residual Plot&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;regression_residuals.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

<span class="c1"># Plot distribution of residuals</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Residuals&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Distribution of Residuals&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;regression_residuals_distribution.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Visualizations saved to output directory&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="working-with-different-sequence-types">Working with Different Sequence Types</h2>
<p><code>fast-seqfunc</code> automatically detects and handles different types of biological sequences:</p>
<ul>
<li>DNA (containing A, C, G, T)</li>
<li>RNA (containing A, C, G, U)</li>
<li>Proteins (containing amino acid letters)</li>
</ul>
<div class="highlight"><pre><span></span><code><span class="c1"># Example with protein sequences</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">fast_seqfunc</span><span class="w"> </span><span class="kn">import</span> <span class="n">generate_random_sequences</span>

<span class="c1"># Generate random protein sequences</span>
<span class="n">protein_sequences</span> <span class="o">=</span> <span class="n">generate_random_sequences</span><span class="p">(</span>
    <span class="n">length</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">count</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">alphabet</span><span class="o">=</span><span class="s2">&quot;ACDEFGHIKLMNPQRSTVWY&quot;</span><span class="p">,</span>  <span class="c1"># Protein alphabet</span>
    <span class="n">fixed_length</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Create a dummy function (e.g., number of hydrophobic residues)</span>
<span class="n">hydrophobic</span> <span class="o">=</span> <span class="s2">&quot;AVILMFYW&quot;</span>
<span class="n">function_values</span> <span class="o">=</span> <span class="p">[</span>
    <span class="nb">sum</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">hydrophobic</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">protein_sequences</span>
<span class="p">]</span>

<span class="c1"># Create dataset</span>
<span class="n">protein_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s2">&quot;sequence&quot;</span><span class="p">:</span> <span class="n">protein_sequences</span><span class="p">,</span>
    <span class="s2">&quot;function&quot;</span><span class="p">:</span> <span class="n">function_values</span>
<span class="p">})</span>

<span class="c1"># Now you could train a model on this protein data using the same workflow</span>
</code></pre></div>
<h2 id="advanced-model-training-options">Advanced Model Training Options</h2>
<p><code>fast-seqfunc</code> uses PyCaret behind the scenes, which allows for customization:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Example with more options</span>
<span class="n">advanced_model_info</span> <span class="o">=</span> <span class="n">train_model</span><span class="p">(</span>
    <span class="n">train_data</span><span class="o">=</span><span class="n">train_data</span><span class="p">,</span>
    <span class="n">test_data</span><span class="o">=</span><span class="n">test_data</span><span class="p">,</span>
    <span class="n">sequence_col</span><span class="o">=</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span>
    <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;function&quot;</span><span class="p">,</span>
    <span class="n">embedding_method</span><span class="o">=</span><span class="s2">&quot;one-hot&quot;</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;regression&quot;</span><span class="p">,</span>
    <span class="n">optimization_metric</span><span class="o">=</span><span class="s2">&quot;r2&quot;</span><span class="p">,</span>
    <span class="c1"># Additional PyCaret setup options:</span>
    <span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>  <span class="c1"># Use all available CPU cores</span>
    <span class="n">fold</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>     <span class="c1"># 5-fold cross-validation</span>
    <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Normalize features</span>
    <span class="n">polynomial_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Generate polynomial features</span>
    <span class="n">feature_selection</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Perform feature selection</span>
<span class="p">)</span>
</code></pre></div>
<h2 id="loading-and-reusing-models">Loading and Reusing Models</h2>
<p>You can load saved models for reuse:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Load a previously saved model</span>
<span class="n">loaded_model_info</span> <span class="o">=</span> <span class="n">load_model</span><span class="p">(</span><span class="n">model_path</span><span class="p">)</span>

<span class="c1"># Use the loaded model for predictions</span>
<span class="n">new_predictions</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">loaded_model_info</span><span class="p">,</span> <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">])</span>

<span class="c1"># Verify the predictions match those from the original model</span>
<span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">new_predictions</span><span class="p">)</span>
</code></pre></div>
<h2 id="conclusion">Conclusion</h2>
<p>You've now learned how to:
1. Prepare sequence-function data
2. Train a regression model using <code>fast-seqfunc</code>
3. Make predictions on new sequences
4. Evaluate model performance
5. Visualize results</p>
<p>For more advanced features and applications, check out the <a href="../../api_reference/">API reference</a> and <a href="../classification_tutorial/">additional tutorials</a>.</p>
<h2 id="next-steps">Next Steps</h2>
<ul>
<li>Try different regression tasks (e.g., "motif_position", "interaction")</li>
<li>Experiment with different model parameters</li>
<li>Apply these techniques to your own sequence-function data</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/ericmjl" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["instant", "tabs"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../config.js"></script>
      
    
  </body>
</html>